photon-client/src/main/scala/com/linkedin/photon/ml/hyperparameter/VectorRescaling.scala

```diff
@@ -0,0 +1,127 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.hyperparameter
+
+import breeze.linalg.DenseVector
+
+import com.linkedin.photon.ml.util.DoubleRange
+
+/**
+ * Functions to transform/scale forward/backward hyper-parameters in vectors
+ */
+object VectorRescaling {
+
+  /**
+   * Apply forward transformation to a subset of elements in a vector.
+   *
+   * @param vector A DenseVector.
+   * @param transformMap A Map with key-value pairs of indices and names of transform functions.
+   * @return The transformed vector.
+   */
+  def transformForward(
+      vector: DenseVector[Double],
+      transformMap: Map[Int, String]): DenseVector[Double] = {
+
+    val vectorTransformed = vector.copy
+
+    transformMap.foreach { case (index, transform) =>
+      transform match {
+        case "LOG" => vectorTransformed(index) = Math.log10(vectorTransformed(index))
+        case "SQRT" => vectorTransformed(index) = Math.sqrt(vectorTransformed(index))
+      }
+    }
+
+    vectorTransformed
+  }
+
+  /**
+   * Apply backward transformation to a subset of elements in a vector.
+   *
+   * @param vector A DenseVector.
+   * @param transformMap A Map with key-value pairs of indices and names of transform functions.
+   * @return The transformed vector.
+   */
+  def transformBackward(
+      vector: DenseVector[Double],
+      transformMap: Map[Int, String]): DenseVector[Double] = {
+
+    val vectorTransformed = vector.copy
+
+    transformMap.foreach { case (index, transform) =>
+      transform match {
+        case "LOG" => vectorTransformed(index) = Math.pow(10, vectorTransformed(index))
+        case "SQRT" => vectorTransformed(index) = Math.pow(vectorTransformed(index), 2)
+      }
+    }
+
+    vectorTransformed
+  }
+
+  /**
+   * Apply forward scaling to a vector. Given a range [a, b] and an element x of the vector,
+   * y = (x - a) / (b - a), if x is continuous;
+   * y = (x - a) / (b - a + 1), if x is discrete.
+   *
+   * @param vector A DenseVector.
+   * @param ranges A sequence of ranges for every element in the vector to be scaled.
+   * @param discreteIndexSet A Set with indices of discrete elements in the vector.
+   * @return The scaled vector.
+   */
+  def scaleForward(
+      vector: DenseVector[Double],
+      ranges: Seq[DoubleRange],
+      discreteIndexSet: Set[Int] = Set()): DenseVector[Double] = {
+
+    val vectorScaled = vector.copy
+
+    val start = DenseVector(ranges.map(_.start).toArray)
+    val end = DenseVector(ranges.map(_.end).toArray)
+    val discreteAdj = DenseVector.tabulate(ranges.length) { i =>
+      if (discreteIndexSet.contains(i)) 1.0 else 0.0
+    }
+    vectorScaled :-= start
```

Comment:
Minor: this is a subjective thing, but I feel there should be an empty line above `vectorScaled :-= start`, to create more of a visual separation between this and the declarations above.